Async wallet scanner

j-berman opened the pull request:
Ready for review.

I've observed the following speed-ups benchmarking versus wallet2:

| Setup | % faster than wallet2 |
| --- | --- |
| Remote node, clearnet | **50-60%** faster |
| Remote node, tor | **40-45%** faster |
| Local node | **25-35%** faster |

Benchmark results can vary widely based on setup (client/node internet speeds and machines). [Source for the scanner used to benchmark](https://github.com/j-berman/monero/commit/9c18795619236ac1cc8c5f71deab6a6987c38066).

### How it works

Assume the user's wallet must start scanning blocks from height 5000.

1. The scanner begins by launching 10 RPC requests in parallel to fetch chunks of blocks as follows:

```
request 0: { start_height: 5000, max_block_count: 20 }
request 1: { start_height: 5020, max_block_count: 20 }
...
request 9: { start_height: 5180, max_block_count: 20 }
```

2. As soon as any single request completes, the wallet immediately parses that chunk.
	- This is all in parallel. For example, as soon as request 7 responds, the wallet immediately begins parsing that chunk in parallel to any other chunks it's already parsing.
3. If a chunk does not include a total of `max_block_count` blocks, and the chunk is not the tip of the chain, this means there was a "gap" in the chunk request. The scanner launches another parallel RPC request to fill in the gap.
    - This gap can occur because the server will **not** return a chunk of blocks greater than [100mb](https://github.com/monero-project/monero/blob/053ba2cf07649cea8134f8a188685ab7a5365e5c/src/cryptonote_core/blockchain.cpp#L65) or [20,000 txs](https://github.com/monero-project/monero/blob/053ba2cf07649cea8134f8a188685ab7a5365e5c/src/cryptonote_config.h#L128) via the `/getblocks.bin` RPC endpoint.
    - The gap is from `(req.start_height + res.blocks.size())` to `(req.start_height + req.max_block_count)`.
4. As soon as the scanner finishes parsing the chunk, it immediately submits another parallel RPC request.
5. In parallel, the scanner identifies a user's received (and spent) enotes in each chunk.
	 - For example, as soon as request 7 responds and the wallet parses it, the wallet scans that chunk in parallel to any other chunks it's already scanning.
6. Once a single chunk is fully scanned locally, the scanner launches a parallel task to fetch and scan the next chunk.
7. Once the scanner reaches the tip of the chain (the terminal chunk), the scanner terminates.

### Some technical highlights

- The wallet scanner in this PR is backwards compatible with existing daemons (though it would need to point to a daemon running the updates in this PR to realize the perf speed-up).
- On error cases such as the daemon going offline, the same wallet errors that wallet2 uses (that the wallet API expects) are propagated up to the higher-level Seraphis lib.
- The implementation in this PR uses an http client connection pool (reusing the epee http client) to support parallel network requests ([related](https://github.com/seraphis-migration/wallet3/issues/58)).
- A developer using the scanner can "bring their own blocks/network implementation" to the scanner by providing a callback function of the following type as a param to the async scanner constructor: `std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response)>`.
	- Thanks to this approach, in relatively short order, I could test swapping the epee http client for libcurl's easy interface ([souce](https://github.com/j-berman/monero/commit/d814a3c077ebdacf24caf2913e96224abccb7160)).

### Major TODO's for this PR

- ~~Cleanup the formatting, code structure, code names and commit history.~~
    - DONE
- ~~Implement backwards compatibility so the scanner will work pointing to nodes running today.~~
    - DONE: also implemented multithreaded scanning so perf won't drop if pointing to a node not running this PR.
- ~~Add unit tests.~~
    - DONE:
        - Added functional tests that the async scanner can scan the chain and identify various legacy txs constructed by wallet2.
        - Added core tests that scanning util can parse all legacy enote versions from different legacy tx types.
            - It would be nice to test that the scanner can fully scan all legacy version enotes (and decode amounts).
- ~~Try to smooth out the scanner's progress.~~
    - ~~The scanner completes scanning many chunks in intermittent waves and then sits waiting (uncomfortably) until the next wave of chunks; it doesn't scan in a steady stream.~~
    - DONE: reduced the `max_chunk_size_hint` from 1000 to 20 to fetch chunks 20 blocks at a time. Overall time to scan remains the same (potentially slightly faster) and feedback to the user on scanner progress is significantly smoother.

### Other TODO's for this PR

- Pre-RCT enotes can be scanned by the Seraphis wallet lib, however, the Seraphis wallet lib does not scan enough data to be able to spend them via legacy wallet2 today (without an RPC request that would reveal the enote to the daemon).
    - The Seraphis wallet lib would at minimum need to keep track of the global output ID by amount in order to capture enough data to pass to wallet2 to spend a pre-RCT enote before the Seraphis fork.
        - The Seraphis wallet lib would need to keep track of the global output ID by amount forever if pre-RCT outputs must be spent in separate rings from RCT outputs ([related](https://github.com/monero-project/research-lab/issues/59#issuecomment-1778565614)).
  - UPDATE: this is being worked on here https://github.com/UkoeHB/monero/pull/40
- ~~Move "mock" implementations to the "production-ready" sections they belong.~~
    - DONE: moved production-ready implementations into expected sections. Mocks that are not ready for production are kept as mocks.
- ~~Complete TODO's in the code.~~
    - DONE: TODO's necessary an initial PR are done.
- ~~As soon as the scanner encounters an error or encounters the terminal chunk, it should immediately cancel any pending tasks it's waiting on and error out quickly.~~
  - DONE: small chunk requests (1000 blocks at a time -> 20 blocks at a time) took care of this, the async scanner cancels pending tasks and errors out quickly.
- ~~Remove the blockchain scanner utility from this PR (and all changes to existing wallet functions for this utility). It's there to help see how the scanner works.~~
    - DONE: the functional test now demonstrates how the scanner works)

### Misc.

There are a few things from this PR I can PR to the Monero repo today in bite-sized PR's:

- Moving wallet error logic out of `wallet2` and into `wallet/wallet_errors.h`. I did this so I could reuse wallet2's error logic in this PR, which the higher level wallet API also expects.
- Modifying the `get_blocks.bin` daemon RPC endpoint:
	- I added an optional `max_block_count` param to the request (not a breaking change)
	- If the request includes a `start_height` that is *greater than* the chain tip and the request field `fail_on_high_height` is false, then instead of erroring, the server returns empty blocks and the chain height.
	- The response includes an optional top block hash field in addition to the chain height.
		- ~~TODO: make this field optional in the response so it's not a breaking change.~~ (DONE)
- A new functional test for direct wallet2 <> live RPC daemon interactions (and a simple daemon RPC client helper that the test uses).

UkoeHB commented:
Nice to see this coming along :)

jeffro256 commented on src/seraphis_mocks/mock_http_client_pool.h:
```diff
@@ -0,0 +1,131 @@
+private:
+    const std::string m_daemon_address;
+    const boost::optional<epee::net_utils::http::login> m_daemon_login;
+    const epee::net_utils::ssl_options_t m_ssl_support;
```
It would be good to put proxy support here

j-berman replied:
Yep, will do

jeffro256 commented on src/seraphis_mocks/mock_http_client_pool.h:
```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2022, The Monero Project
```
These http_client_pool classes can probably go in the src/net folder, yeah?

j-berman replied:
Agree, I'll throw them in there once it's in a ready-for-use state

jeffro256 commented on src/seraphis_mocks/mock_http_client_pool.cpp:
```diff
@@ -0,0 +1,91 @@
+//-------------------------------------------------------------------------------------------------------------------
+void ClientConnectionPool::release_http_client(size_t http_client_index)
+{
+    // Make the connection available for use again
```
Will we ever prune the list of HTTP clients?

j-berman replied:
I think it makes sense that once the scanner finishes a full pass, only 1 connection should be kept open, the rest can be closed. Seems good housekeeping.

In the future, I think this connection pool could potentially be a good class to build multi-daemon functionality with (1 client -> many different daemons), in which case it may make sense to keep connections to different daemons open. But that's probably a discussion for another time, just putting the thought out since I know you were interested in pushing that architecture forward

jeffro256 commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,174 @@
+    std::unique_ptr<sp::scanning::LedgerChunk> get_onchain_chunk();
+
+    /// stop the current scanning process (should be no-throw no-fail)
+    void terminate_scanning() override { /* no-op */ }
```
Are there plans to make it stoppable in the future?

j-berman replied:
If the wallet is scanning and the user changes their restore height or wants to use a different wallet, that should cleanly terminate the scanner

UkoeHB replied:
If stopping from outside is needed, it can be easily implemented with an atomic signal. Not needed for the initial implementation.

jeffro256 commented on src/seraphis_mocks/enote_finding_context_mocks.h:
```diff
@@ -102,6 +104,51 @@ class EnoteFindingContextLedgerMockLegacy final : public EnoteFindingContextLedg
+        std::list<sp::ContextualBasicRecordVariant> &collected_records) const;
+
+    /// abstracted function that gets blocks via RPC request
+    const std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request&, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response&)> &rpc_get_blocks;
```
Should this be marked `private`?

j-berman replied:
I'm about to take a deeper pass at class structure and will probably make a lot of changes in line with this comment. Worth re-visiting after making those changes

j-berman commented:
Probably worth holding off on a deeper review until it's ready, almost there :)

j-berman committed:
wallet2: make error handling functions reusable

Co-authored-by: jeffro256 <jeffro256@tutanota.com>

j-berman committed:
daemon RPC: add max_block_count field to getblocks.bin request

j-berman committed:
daemon RPC: getblocks.bin return chain's top_block_hash

j-berman committed:
daemon RPC: optional don't error getblocks.bin when req.start_height > height

j-berman committed:
tools: simple daemon rpc client

Useful for tests, similar API to utils/python-rpc/framework/daemon.py

j-berman committed:
functional tests: transfers & scanning direct wallet2 -> RPC daemon

Adds a new functional test for direct wallet2 -> live RPC daemon
interactions. This sets up a framework to test pointing the
Seraphis wallet lib to a live daemon.

Tests sending and scanning:
- a normal transfer
- a sweep single (0 change)
- to a single subaddress
- to 3 subaddresses (i.e. scanning using additional pub keys)

j-berman committed:
scan machine: option to force reorg avoidance increment first pass

- when pointing to a daemon that does not support returning empty
blocks when the client requests too high of a height, we have to
be careful in our scanner to always request blocks below the chain
tip, in every request.
- by forcing the reorg avoidance increment on first pass, we make
sure clients will always include the reorg avoidance increment
when requesting blocks from the daemon, so the client can expect
the request for blocks should *always* return an ok height.

j-berman committed:
core tests: check conversion tool on all legacy enote version types

Stil TODO:
- check complete scanning on all enote types
- hit every branch condition for all enote versions

j-berman committed:
conn pool mock: epee http client connection pool

- Enables concurrent network requests using the epee http client.
- Still TODO for production:

1) close_connections
2) require the pool respect max_connections

j-berman committed:
enote finding context: IN LegacyUnscannedChunk, OUT ChunkData

- finds owned enotes by legacy view scanning a chunk of blocks

j-berman committed:
enote finding context: multithreaded view scanning

j-berman committed:
async: function to remove minimum element from token queue

- Useful when we want to remove elements of the token queue in an
order that is different than insertion order.

j-berman committed:
async scanner: scan via RPC, fetching & scanning parallel chunks

*How it works*

Assume the user's wallet must start scanning blocks from height
5000.

1. The scanner begins by launching 10 RPC requests in parallel to
fetch chunks of blocks as follows:

```
request 0: { start_height: 5000, max_block_count: 20 }
request 1: { start_height: 5020, max_block_count: 20 }
...
request 9: { start_height: 5180, max_block_count: 20 }
```

2. As soon as any single request completes, the wallet immediately
parses that chunk.
  - This is all in parallel. For example, as soon as request 7
  responds, the wallet immediately begins parsing that chunk in
  parallel to any other chunks it's already parsing.
3. If a chunk does not include a total of max_block_count blocks,
and the chunk is not the tip of the chain, this means there was a
"gap" in the chunk request. The scanner launches another parallel
RPC request to fill in the gap.
  - This gap can occur because the server will **not** return a
  chunk of blocks greater than 100mb (or 20k txs) via the
  /getblocks.bin` RPC endpoint
  ([`FIND_BLOCKCHAIN_SUPPLEMENT_MAX_SIZE`](https://github.com/monero-project/monero/blob/053ba2cf07649cea8134f8a188685ab7a5365e5c/src/cryptonote_core/blockchain.cpp#L65))
  - The gap is from `(req.start_height + res.blocks.size())` to
  `(req.start_height + req.max_block_count)`.
4. As soon as the scanner finishes parsing the chunk, it
immediately submits another parallel RPC request.
5. In parallel, the scanner identifies a user's received (and
spent) enotes in each chunk.
  - For example, as soon as request 7 responds and the wallet
  parses it, the wallet scans that chunk in parallel to any other
  chunks it's already scanning.
6. Once a single chunk is fully scanned locally, the scanner
launches a parallel task to fetch and scan the next chunk.
7. Once the scanner reaches the tip of the chain (the terminal
chunk), the scanner terminates.

*Some technical highlights*

- The wallet scanner is backwards compatible with existing daemons
(though it would need to point to an updated daemon to realize the
perf speed-up).
- On error cases such as the daemon going offline, the same wallet
errors that wallet2 uses (that the wallet API expects) are
propagated up to the higher-level Seraphis lib.
- The implementation uses an http client connection pool (reusing
the epee http client) to support parallel network requests
([related](https://github.com/seraphis-migration/wallet3/issues/58)).
- A developer using the scanner can "bring their own blocks/network
implementation" to the scanner by providing a callback function of
the following type as a param to the async scanner constructor:
`std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response)>`

j-berman committed:
functional tests: async scanner config for old daemon version

j-berman commented:
Ready for review :)

Noting there are 2 significant aspects of this scanner that will need to be dealt with. I think it'll be easier to handle in follow-up PR's.

**1. The subaddress lookahead.**

- The `EnoteFindingContextMockLegacy` takes as input a constant `legacy_subaddress_map` to see enotes received to subaddresses.
	- This means all subaddresses the async scanner will scan must be known to the wallet **before it starts scanning**.
- However, wallet2 currently implements a lookahead that expands when scanning tx outputs serially.
	- If wallet2 identifies a received output to a fresh subaddress, it then expands the lookahead to identify subsequent outputs received to higher indexed subaddresses.
	- This means the async scanner can miss txs to a subaddress that wallet2 would pick up.
- In order to match wallet2's behavior with max performance, a `LegacyBasicEnoteRecord` could instead be view tag matched enotes and their `nominal_spendkey`, rather than known-owned view scanned enotes.
	- As the scanner then consumes a chunk of `LegacyBasicEnoteRecord`'s serially in `process_chunk_full_legacy` and identifies owned enotes, it then expands the suabddress lookahead as needed.

**2. Pool handling.**

- The async scanner does not deal with pool handling in this PR.
- The Seraphis lib scanner does a "scanning pass" in this order: 1) scan chain, 2) scan pool, 3) scan chain again in case the chain advanced.
- wallet2 does pool handling through the same single RPC request as fetching blocks from the daemon RPC (`/getblocks.bin`) in order to reduce round trips to the daemon.
- The async scanner design could change to optimally handle pool txs in line with the way wallet2 currently does it. Some scratchpad thoughts:
	- I think it would make sense to re-structure the Seraphis lib to simply do a single "scanning pass" (scan chain then pool).
	- The Seraphis lib currently assumes it's finished scanning upon encountering an empty chunk. The async scanner could pass an empty chunk back once all pool txs have been fetched.
	- Alternatively, the `AsyncScanContextLegacy` could cache pool txs and then that cache can be reused for `get_nonledger_chunk`.

j-berman commented:
To reviewers: I suggest going commit-by-commit and working up to the final async scanner.

10/13 of the commits are theoretically independent and can be PR'd separately.

If the design generally looks solid, I can divide it up into smaller PR's to ease review. Additionally we can PR the daemon RPC changes to monerod at any time, so that daemons would be prepared for this scanner well in advance of releasing the Seraphis lib in a functional wallet to end-users.

UkoeHB commented:
> I think it would make sense to re-structure the Seraphis lib to simply do a single "scanning pass" (scan chain then pool).

I intentionally added a follow-up pass because there is a race condition between chain scanning and pool scanning that can cause mined blocks to contain txs invisible to a wallet even after it claims to have scanned the pool.

j-berman commented:
Opened a separate issue to discuss since I don't think it needs to affect this PR: https://github.com/UkoeHB/monero/issues/41

DangerousFreedom1984 commented on src/wallet/wallet2.h:
```diff
@@ -74,15 +74,6 @@
 #undef MONERO_DEFAULT_LOG_CATEGORY
 #define MONERO_DEFAULT_LOG_CATEGORY "wallet.wallet2"
 
-#define THROW_ON_RPC_RESPONSE_ERROR(r, error, res, method, ...) \
```
I agree that a better place for that is at `wallet_errors.h`.

DangerousFreedom1984 commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -176,15 +189,20 @@ namespace cryptonote
+      bool        fail_on_high_height;
       uint64_t    pool_info_since;
+      uint64_t    max_block_count;
+
```
A noob question: you are only bumping the CORE_RPC_VERSION_MINOR but you are changing these structures, how do you know it is still compatible with other major versions? Because you are serializing as an OPTIONAL with a default value?

j-berman replied:
This was the intent, but thinking on it further I'd feel more comfortable if behavior works from default falsey values. I'm going to modify these changes slightly based on this comment.

j-berman replied:
Done in the latest

DangerousFreedom1984 commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -2765,16 +2765,19 @@ bool Blockchain::find_blockchain_supplement(const std::list<crypto::hash>& qbloc
   LOG_PRINT_L3("Blockchain::" << __func__);
   CRITICAL_REGION_LOCAL(m_blockchain_lock);
 
   // if a specific start height has been requested
```
Wouldnt it be better to have this whole block like this:
```
  // if a specific start height has been requested
  uint64_t top_height;
  top_hash = m_db->top_block_hash(&top_height);
  if(req_start_block > 0)
  {
    // if requested height is higher than our chain, return false -- we can't help
    total_height = top_height + 1;
    if (req_start_block >= total_height)
    {
      return false;
    }
    start_height = req_start_block;
  }
  else
  {
    if(!find_blockchain_supplement(qblock_ids, start_height))
    {
      return false;
    }
  }

  db_rtxn_guard rtxn_guard(m_db);
  total_height = top_height + 1;
```
So we avoid calling `top_block_hash` twice in some cases?

j-berman replied:
The `db_rtxn_guard` seems to be there to guarantee a consistent view of the db. Even though the `m_blockchain_lock` is held, which *should* prevent the db from changing between the 2 db reads for the top block hash, it looks as though the 2 distinct db reads (1 before and 1 after the read txn guard) are there to protect against that scenario anyway (the blockchain changing between the 2 calls).

I would personally prefer to maintain the existing concurrency model and number of db calls for ease of review here.

DangerousFreedom1984 commented on src/common/rpc_client.h:
```diff
@@ -148,4 +187,75 @@ namespace tools
+
+    cryptonote::COMMAND_RPC_GET_TRANSACTIONS::response get_transactions(const std::vector<std::string> &txs_hashes)
+    {
+      cryptonote::COMMAND_RPC_GET_TRANSACTIONS::request req = AUTO_VAL_INIT(req);
+      cryptonote::COMMAND_RPC_GET_TRANSACTIONS::response res = AUTO_VAL_INIT(res);
+      req.txs_hashes = txs_hashes;
+      req.decode_as_json = false;
+      CHECK_AND_ASSERT_THROW_MES(m_rpc_client.basic_rpc_request(req, res, "/get_transactions"), "failed to get transactions");
+      return res;
+    }
+
+    cryptonote::COMMAND_RPC_FLUSH_TRANSACTION_POOL::response flush_txpool()
+    {
+      cryptonote::COMMAND_RPC_FLUSH_TRANSACTION_POOL::request req = AUTO_VAL_INIT(req);
+      cryptonote::COMMAND_RPC_FLUSH_TRANSACTION_POOL::response res = AUTO_VAL_INIT(res);
+      CHECK_AND_ASSERT_THROW_MES(m_rpc_client.basic_json_rpc_request(req, res, "flush_txpool"), "failed to flush txpool");
+      return res;
+    }
+
+    cryptonote::COMMAND_RPC_GENERATEBLOCKS::response generateblocks(const std::string &address, uint64_t amount_of_blocks)
+    {
+      cryptonote::COMMAND_RPC_GENERATEBLOCKS::request req = AUTO_VAL_INIT(req);
+      cryptonote::COMMAND_RPC_GENERATEBLOCKS::response res = AUTO_VAL_INIT(res);
+      req.amount_of_blocks = amount_of_blocks;
+      req.wallet_address = address;
+      req.prev_block = "";
+      req.starting_nonce = 0;
+      CHECK_AND_ASSERT_THROW_MES(m_rpc_client.basic_json_rpc_request(req, res, "generateblocks"), "failed to generate blocks");
+      return res;
+    }
+
+    cryptonote::COMMAND_RPC_GET_LAST_BLOCK_HEADER::response get_last_block_header()
+    {
+      cryptonote::COMMAND_RPC_GET_LAST_BLOCK_HEADER::request req = AUTO_VAL_INIT(req);
+      cryptonote::COMMAND_RPC_GET_LAST_BLOCK_HEADER::response res = AUTO_VAL_INIT(res);
+      req.client = "";
+      CHECK_AND_ASSERT_THROW_MES(m_rpc_client.basic_json_rpc_request(req, res, "get_last_block_header"), "failed to get last block header");
+      return res;
+    }
+  };
```
Why is this whole block necessary? Do the scanner needs to create its own new set of RPC functions to interact with the blockchain?

j-berman replied:
These are boilerplate helper functions to construct the RPC requests in the functional tests. Grep `daemon.generateblocks` to see how it's used in both the `.py` tests and `wallet_scanner.cpp` as an example.

DangerousFreedom1984 commented on src/seraphis_main/enote_finding_context.cpp:
```diff
@@ -0,0 +1,191 @@
+    }
+}
+//-------------------------------------------------------------------------------------------------------------------
+void EnoteFindingContextLegacyMultithreaded::view_scan_chunk(const LegacyUnscannedChunk &legacy_unscanned_chunk,
```
Nice! How the performance of these two `view_scan_chunk` compare?

j-berman replied:
On my machine with 8 threads and a pending chunk queue of size 10, there is no benefit to using this multithreaded enote finding context in the async scanner. My CPU is already maxxed out.

This enote finding context should only be useful when the pending chunk queue size is < number of threads available on the machine. Example scenario: when pointing the scanner to a daemon not running the changes from this PR, it's more efficient to use a pending chunk queue size of 1, and to use this multithreaded enote finding context, since the daemon does not yet know about the `max_block_count` request param. The perf gain for client-side scanning in this case is then sped up by a multiple of the number of threads on the machine.

DangerousFreedom1984 reviewed (commented):
I still have to go through the `scan_context_async_mock` and maybe play with it in some unit_test to better understand and get used to it. Will do soon.

Awesome work! Thank you!

UkoeHB commented on src/cryptonote_core/blockchain.cpp:
```diff
@@ -2765,16 +2765,19 @@ bool Blockchain::find_blockchain_supplement(const std::list<crypto::hash>& qbloc
   CRITICAL_REGION_LOCAL(m_blockchain_lock);
 
   // if a specific start height has been requested
+  uint64_t top_height;
```
I'd move `top_height` to the places where it is set, I found this code pretty confusing at first.

UkoeHB commented on src/rpc/core_rpc_server.cpp:
```diff
@@ -702,26 +702,35 @@ namespace cryptonote
         {
           res.start_height = 0;
           res.current_height = last_block_height + 1;
+          res.top_block_hash = std::move(last_block_hash);
```
```suggestion
          res.top_block_hash = last_block_hash;
```
`crypto::hash` is not allocated.

UkoeHB commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -88,7 +101,7 @@ namespace cryptonote
 // Don't go over 32767 for any of these
 #define CORE_RPC_VERSION_MAJOR 3
-#define CORE_RPC_VERSION_MINOR 13
+#define CORE_RPC_VERSION_MINOR 14
```
I'd undo this minor version change, since it will cause rebase conflicts with core. The minor version tracking is a big PITA.

If you PR your changes to monero core, then I can rebase `seraphis_lib` (before or after merging this PR is fine with me).

UkoeHB commented on src/rpc/core_rpc_server_commands_defs.h:
```diff
@@ -244,6 +263,7 @@ namespace cryptonote
         KV_SERIALIZE(blocks)
         KV_SERIALIZE(start_height)
         KV_SERIALIZE(current_height)
+        KV_SERIALIZE_VAL_POD_AS_BLOB_OPT(top_block_hash, crypto::null_hash)
```
Do you have a way of detecting in the scanner if pointing to a daemon that doesn't return top block hashes?

j-berman replied:
The `get_version` RPC endpoint returns the major and minor RPC version: https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/rpc/core_rpc_server.cpp#L2986

Wallets currently hit that endpoint first to determine if a daemon is compatible. Wallets will know which minor version returns top block hashes (and which won't fail the request on high height, and will respect `max_block_count`)

UkoeHB commented on src/seraphis_main/scan_machine.cpp:
```diff
@@ -71,13 +71,14 @@ enum class ContiguityCheckResult : unsigned char
     const std::uint64_t num_reorg_avoidance_backoffs)
 {
     // 1. start at a depth of zero
     // - this allows us to avoid accidentally reorging your data store if the scanning backend only has a portion
     //   of the blocks in your initial reorg avoidance depth range available when 'get chunk' is called (in the case
     //   where there wasn't actually a reorg and the backend is just catching up)
```
```suggestion
    // - this allows us to avoid accidentally reorging your data store if the scanning backend only has a portion
    //   of the blocks in your initial reorg avoidance depth range available when 'get chunk' is called (in the case
    //   where there wasn't actually a reorg and the backend is just catching up)
    // - for example, if your refresh index is 100, your start index is 150, and your avoidance increment is 10, if
    //   you start scanning but your scan contex has only processed blocks 0-145 (out of let's say 200), then if we
    //   include a reorg avoidance backoff for the first processed chunk, then blocks [140, ..) will be requested
    //   from the scan context and only [140, 145] might be returned; when the chunk consumer gets those blocks, it
    //   will think blocks 145-150 were reorged away; a scan context *could* avoid returning a partial chunk if it
    //   knows there are unprocessed blocks in the chunk, but for robustness we can't assume all contexts will do so
```

Sticking this clarification in since it threw me for a bit trying to remember the rationale here.

I recommend adding a comment of your own justifying/explaining `force_reorg_avoidance_increment`.

UkoeHB commented on src/seraphis_main/CMakeLists.txt:
```diff
@@ -59,6 +60,7 @@ monero_add_library(seraphis_main
 target_link_libraries(seraphis_main
   PUBLIC
+    async
     cncrypto
```
```suggestion
  PUBLIC
    cncrypto
```
See comment on `enote_finding_context.h`.

UkoeHB commented on src/seraphis_main/enote_finding_context.h:
```diff
@@ -84,4 +89,131 @@ class EnoteFindingContextLedger
+
+////
+/// LegacyUnscannedChunk: a chunk of blocks ready to be legacy view scanned
+// - the blocks are expected to match their order on chain
```
```suggestion
// - the blocks are expected to match their order on-chain
```

UkoeHB commented on src/seraphis_main/enote_finding_context.h:
```diff
@@ -84,4 +89,131 @@ class EnoteFindingContextLedger
+////
+/// LegacyUnscannedBlock: a block that is ready to be legacy view scanned
+// - the txs are expected to be ordered as they appear in the block, where
+// the first tx is the miner tx
```
```suggestion
// - the txs are expected to be ordered as they appear in the block, where
//   the first tx is the miner tx
```

UkoeHB commented on src/seraphis_main/enote_finding_context.h:
```diff
@@ -84,4 +89,131 @@ class EnoteFindingContextLedger
+/// LegacyUnscannedChunk: a chunk of blocks ready to be legacy view scanned
+// - the blocks are expected to match their order on chain
+///
+typedef std::vector<LegacyUnscannedBlock> LegacyUnscannedChunk;
```
```suggestion
using LegacyUnscannedChunk = std::vector<LegacyUnscannedBlock>;
```
Stylistic consistency.

Using statements are more clear IMO.

UkoeHB commented on src/seraphis_main/enote_finding_context.h:
```diff
@@ -84,4 +89,131 @@ class EnoteFindingContextLedger
+// - find owned enotes from legacy view scanning using actual chain data
+// - scans each tx in a chunk of blocks serially in order
+///
+class EnoteFindingContextLegacySimple final : public EnoteFindingContextLegacy
```
These implementations should be moved to `seraphis_impl`. This file is for the abstract base classes for enote finding contexts.

It might even be fine to move this entire file to `seraphis_impl` (and split between virtual base class file and legacy implementations), since finding contexts are not used anywhere in `seraphis_main` (a left-over from when they were used more directly).

UkoeHB commented on src/seraphis_main/enote_finding_context.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    // 7. set results
+    for (auto &brpt : basic_records_per_tx)
+        chunk_data_out.basic_records_per_tx.emplace(std::move(brpt));
```
Simple but effective, LGTM. Did you do any testing if it performs better with more txs processed per task? I'm not sure what the per-task overhead is.

j-berman replied:
> Did you do any testing if it performs better with more txs processed per task?

Nay, going to punt that for now

UkoeHB commented on src/seraphis_mocks/mock_http_client_pool.h:
```diff
@@ -0,0 +1,139 @@
+
+//forward declarations
+
+namespace sp
```
```suggestion
//forward declarations


namespace sp
```
Tiny style nit

UkoeHB commented on src/seraphis_mocks/mock_http_client_pool.h:
```diff
@@ -0,0 +1,139 @@
+        typename COMMAND_TYPE::response &res)
+    {
+        // Acquire an http client from the connection pool
+        std::size_t http_client_index = acquire_unused_http_client();
```
```suggestion
        std::size_t http_client_index = this->acquire_unused_http_client();
```
Disambiguate (same as `self.` in rust).

UkoeHB commented on src/seraphis_mocks/mock_http_client_pool.h:
```diff
@@ -0,0 +1,139 @@
+        std::size_t http_client_index = acquire_unused_http_client();
+        CHECK_AND_ASSERT_THROW_MES(m_http_client_pool.size() > http_client_index, "http client index is too high");
+        auto scope_exit_handler = epee::misc_utils::create_scope_leave_handler([this, http_client_index]{
+            release_http_client(http_client_index);
```
```suggestion
            this->release_http_client(http_client_index);
```

UkoeHB commented on src/seraphis_mocks/mock_http_client_pool.cpp:
```diff
@@ -0,0 +1,96 @@
+            found_unused_client = true;
+            break;
+        }
+    }
```
```suggestion
    // Check if we have any clients available for use
    for (std::size_t i = 0; i < m_http_client_pool.size(); ++i)
    {
        if (m_http_client_pool[i].in_use)
            continue;

        m_http_client_pool[i].in_use = true;
        return i;
    }
```
Simpler, less nesting here and after. I am a *huge* fan of early-out/early-continue patterns like this.

UkoeHB commented on tests/core_tests/enote_scanning.h:
```diff
@@ -0,0 +1,138 @@
+#pragma once 
+#include "chaingen.h"
+
+struct gen_enote_tx_validation_base : public test_chain_unit_base
```
Normally I would say to put these test changes in a separate PR since they are unrelated to the async wallet scanner, but it's fine in this case.

j-berman replied:
Hah, I actually was about to separate it when finalizing the PR and thought heh, I'm just gonna leave it. I initially wanted to use the scanner in these tests, but that would be a bit of a lift here

UkoeHB commented on tests/functional_tests/functional_tests_rpc.py:
```diff
@@ -146,7 +146,7 @@ def kill():
     print('[TEST STARTED] ' + test)
     sys.stdout.flush()
-    cmd = [python, srcdir + '/' + test + ".py"]
+    cmd = [python, srcdir + '/' + test + ".py"] if test != 'wallet_scanner' else [FUNCTIONAL_TESTS_DIRECTORY + '/functional_tests', '--' + test]
```
Looks a bit jank

UkoeHB commented on tests/functional_tests/wallet_scanner.cpp:
```diff
@@ -0,0 +1,667 @@
+                    sp::mocks::ClientConnectionPool::http_mode::BIN,
+                    "/getblocks.bin",
+                    req,
+                    res);
```
This internally locks a mutex, which means you will have some random contention even though the window under lock is small. I'd add a `todo` to the connection pool to get away from the mutex.

j-berman replied:
How would you design it to move away from the mutex? Seemed the simplest way to implement this `conn_pool.rpc_command` function, or else I figure the caller would need to handle synchronization, which seems unnecessary (it's a nice API! :) )

UkoeHB replied:
I don't know either, but adding a `todo` will document that the perf could theoretically be improved if someone can figure it out.

UkoeHB commented on tests/functional_tests/wallet_scanner.cpp:
```diff
@@ -0,0 +1,667 @@
+    CHECK_AND_ASSERT_THROW_MES(r, "Failed to refresh enote store");
+
+    // Now that we're done scanning, we can close all open connections and keep 1 open for more RPC calls
+    conn_pool.close_connections(1);
```
This is a bit brittle because it assumes this function completely owns all connections in the pool.

Even though this is a test, it's good to demo how to use the API robustly/correctly since this code will be used as a reference.

j-berman replied:
Hm, I wrote it this way because I think the scanner *should* own all connections in the pool. If the client owns n open connections already when starting the scanner, the scanner can reuse those existing connections. If the client owns 0 open connections when starting the scanner, and the scanner opens 1 connection, then the client can reuse that connection in the future.

To harden this code, I could write a wrapper "Scanner" class that basically does what this `scan_chain` function does, and makes the `conn_pool` a private member variable of that class, but:

1) I still think the scanner should own all connections in the pool for the above reasons
2) I still think it would make sense to call this `conn_pool.close_connections(1)` when the scanner finishes a scan pass (maybe you think the caller is the better place to do that?)

Thoughts?

FWIW the reason I included this unimplemented `close_connections` function call was to demo how to use the API

UkoeHB replied:
> Hm, I wrote it this way because I think the scanner should own all connections in the pool.

I think it's fine to want that. The problem I see in this code is you're passing the pool by reference all the way from the top, which means it *could* be shared by who-knows-what. IMO the pool owner should be the one calling `close_connections` when it knows connections are no longer needed. So if necessary, pass by value in and out of top-level owners, then pass by ref down into the scanning code (which only makes requests, without calling `close_connections`).

j-berman replied:
> pass by value in and out of top-level owners

How bout I use a `unique_ptr` instead of this? That way it's clear the caller is the owner + passing in and out seems a bit like an anti-pattern

> IMO the pool owner should be the one calling `close_connections` when it knows connections are no longer needed

Which line would you move `close_connections` to?

UkoeHB replied:
Ok what you can do is 'claim' the connection pool at the top of this function, which creates a no-move/no-copy RAII wrapper that will remove connections and lock the pool when dropped. Creating the wrapper asserts there are no other claims to the pool (implying there are no connections in use).

Then panic if trying to launch a request on an unclaimed pool.

Basically make pool-use scoped.

j-berman replied:
Ok basically did the above and thought more on it.. technically all the tests also assume the daemon and wallets aren't shared / in use by other owners too, but they could be in theory. Separately, I also wanted to restructure the tests into a class so I don't need to pass the daemon / wallet / conection pool through nested calls. So I implemented a wrapper accessor to each of them that checks for ownership of the respective mutex before allowing access

FWIW it's not really perfect as is, since the underlying daemon / wallet / connection pool can still be accessed without going through the accessors. Maybe you have more thoughts on a better approach still

UkoeHB reviewed (changes requested):
Part 1: everything except `scan_context_async_mock.h/.cpp`

It's looking great so far :)

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,229 @@
+        m_enote_finding_context{enote_finding_context},
+        m_scanner_ready{false},
+        m_threadpool{threadpool},
+        rpc_get_blocks{_rpc_get_blocks}
```
```suggestion
            const std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request&, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response&)> &rpc_get_blocks) :
        m_config{config},
        m_enote_finding_context{enote_finding_context},
        m_scanner_ready{false},
        m_threadpool{threadpool},
        rpc_get_blocks{rpc_get_blocks}
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,229 @@
+    ~AsyncScanContextLegacy()
+    {
+        std::lock_guard<std::mutex> lock{m_async_scan_context_mutex};
+        wait_until_pending_queue_clears();
```
```suggestion
        this->wait_until_pending_queue_clears();
```

UkoeHB replied:
Also, for faster shut down in this case you can tell the TokenQueue to shut down, use force_pop to drain all the remaining tasks, and then wait on all the drained tasks. You'll have to force pop & wait in a loop in case a popped task somehow shoves another task into the queue.

j-berman replied:
I figure we'd want a faster shut down every spot this `wait_until_pending_queue_clears` is called, so I implemented `force_pop` in `wait_until_pending_queue_clears` and added a `TokenQueue::reset()`function to enable flipping `m_is_shutting_down` back to true.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    /// Kick off the scanner starting from the provided index
+    void begin_scanning_from_index(const std::uint64_t start_index, const std::uint64_t max_chunk_size_hint);
```
```suggestion
    void begin_scanning_from_index(const std::uint64_t start_index, const std::uint64_t max_chunk_size_hint) override;
```

Make it clear this is implementing `ScanContextLedger`.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,229 @@
+    void begin_scanning_from_index(const std::uint64_t start_index, const std::uint64_t max_chunk_size_hint);
+
+    /// Get the next chunk from the scanner. Must call begin_scanning_from_index once before get_onchain_chunk
+    std::unique_ptr<sp::scanning::LedgerChunk> get_onchain_chunk();
```
```suggestion
    std::unique_ptr<sp::scanning::LedgerChunk> get_onchain_chunk() override;
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,229 @@
+private:
+    /// config options
+    const AsyncScanContextLegacyConfig &m_config;
+    std::uint64_t m_max_chunk_size_hint;
```
```suggestion
```
Not initialized

j-berman replied:
It's set in `start_scanner`. I gave it a default value in the latest

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static void rpc_get_blocks_internal(const ChunkRequest &chunk_request,
+    const std::function<bool(const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request&, cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response&)> &rpc_get_blocks,
```
```suggestion
    const std::function<bool(
        const cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::request&,
        cryptonote::COMMAND_RPC_GET_BLOCKS_FAST::response&
    )> &rpc_get_blocks,
```
Just noticed it now, but this callback is way outside max width.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    {
+        // must have requested the tip of the chain
+        chunk_context_out.prefix_block_id = rct::hash2rct(res.top_block_hash);
+        chunk_context_out.start_index = res.current_height;
```
Is this 'top block index + 1'?

j-berman replied:
Yep, added a comment

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    for (std::size_t block_idx = 0; block_idx < num_blocks; ++block_idx)
+    {
+        auto &unscanned_block = unscanned_chunk_out[block_idx];
+        unscanned_block.unscanned_txs.resize(1 + res.blocks[block_idx].txs.size());
```
```suggestion
        const auto &res_block = res.blocks[block_idx];
        auto &unscanned_block = unscanned_chunk_out[block_idx];
        unscanned_block.unscanned_txs.resize(1 + res_block.txs.size());
```

Either this or write a separate function for the loop body (imo it's big enough for a new function).

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+
+        prepare_unscanned_legacy_transaction(miner_tx_hash,
+            block.miner_tx,
+            get_total_output_count_before_tx(res.output_indices[block_idx].indices[0].indices),
```
The `.indices[0]` accesses here and below are unchecked afaict.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    const sp::scanning::ChunkContext &chunk_context)
+{
+    if (!chunk_is_terminal_chunk)
+    {
```
```suggestion
    if (chunk_is_terminal_chunk)
        return;

```
Less nesting

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    {
+        // If chunk was smaller than requested, will need to fill the gap
+        const std::size_t chunk_size = sp::scanning::chunk_size(chunk_context);
+        const std::uint64_t gap = requested_chunk_size - chunk_size;
```
```suggestion
        const std::uint64_t gap = math::saturating_sub(requested_chunk_size, chunk_size, 0);
```

j-berman replied:
Underflow shouldn't be possible here because of this in `parse_rpc_get_blocks`:

```
const std::uint64_t num_blocks = std::min((std::uint64_t)res.blocks.size(), chunk_request.requested_chunk_size);
unscanned_chunk_out.resize(num_blocks);
```

In the latest I throw if `requested_chunk_size < chunk_size` because otherwise the rest of the scanner will break if it actually saturates

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+                        .requested_chunk_size = gap
+                    };
+
+                m_pending_chunk_queue.force_push(launch_chunk_task(next_chunk_request, lock));
```
```suggestion
                m_pending_chunk_queue.force_push(this->launch_chunk_task(next_chunk_request, lock));
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    {
+        m_end_scan_index = num_blocks_in_chain;
+        MDEBUG("Set m_end_scan_index: " << m_end_scan_index);
+    }
```
Why is this set only once, and what is the difference from `m_num_blocks_in_chain`?

j-berman replied:
I thought on this some more. The purpose of `m_end_scan_index` was to handle the edge case of the chain advancing while scanning, but it can be done more simply just with `m_num_blocks_in_chain`. Removed `m_end_scan_index` in the latest

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    {
+        m_num_blocks_in_chain = num_blocks_in_chain;
+        MDEBUG("Updated num_blocks_in_chain: " << m_num_blocks_in_chain);
+    }
```
These conditionals need a detailed comment explaining them, like the one on the conditional farther down.

j-berman replied:
Good in the latest?

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        chunk_is_terminal_chunk_out);
+
+    // Check if the chunk was smaller than requested and fill gap if needed
+    fill_gap_if_needed(chunk_is_terminal_chunk_out,
```
```suggestion
    this->update_chain_state(chunk_context_out,
        res.current_height,
        res.top_block_hash,
        chunk_is_terminal_chunk_out);

    // Check if the chunk was smaller than requested and fill gap if needed
    this->fill_gap_if_needed(chunk_is_terminal_chunk_out,
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    bool chunk_is_terminal_chunk = false;
+    try
+    {
+        handle_chunk_context(chunk_request,
```
```suggestion
        this->handle_chunk_context(chunk_request,
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        return boost::none;
+
+    // launch the next task if we expect more and the queue has room
+    launch_next_task_if_room(chunk_is_terminal_chunk);
```
```suggestion
    this->launch_next_task_if_room(chunk_is_terminal_chunk);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+
+    MDEBUG("Finished scanning chunk starting at " << chunk_request.start_index);
+
+    launch_next_task_if_room(chunk_is_terminal_chunk);
```
```suggestion
    this->launch_next_task_if_room(chunk_is_terminal_chunk);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    // find-received-scan raw data
+    // - note: process chunk data can 'do nothing' if the chunk is empty (i.e. don't launch any tasks)
+    sp::scanning::ChunkData chunk_data;
+    m_enote_finding_context.view_scan_chunk(unscanned_chunk, chunk_data);
```
This call is potentially UB because there is no contract on the finding context (which is a non-const reference).

You should change the enote finding context API to be const, and use `mutable` for the threadpool ref in the multithreaded one. Then save the finding context as `const &` in this class.

j-berman replied:
> use mutable for the threadpool ref in the multithreaded one

Can't use mutable on a ref. Fine to make the finding context const and `view_scan_chunk` const, and not make `m_threadpool` mutable? My compiler is ok with it

UkoeHB replied:
Huh guess that's ok. Didn't realize you can access a mutable reference via const `this` ptr.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,229 @@
+    ~AsyncScanContextLegacy()
+    {
+        std::lock_guard<std::mutex> lock{m_async_scan_context_mutex};
+        wait_until_pending_queue_clears();
```
You need to lock this struct to no-move/no-copy, otherwise this destructor does not prevent UB since the reference member variables can be destroyed before this destructor is called.

```cpp
    /// disable copy/move (this is a scoped manager [reference wrapper])
    AsyncScanContextLegacy& operator=(AsyncScanContextLegacy&&) = delete;
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    async::join_signal_t context_join_signal                        = m_threadpool.make_join_signal();
+    async::join_signal_t data_join_signal                           = m_threadpool.make_join_signal();
+    async::join_token_t context_join_token                          = m_threadpool.get_join_token(context_join_signal);
+    async::join_token_t data_join_token                             = m_threadpool.get_join_token(data_join_signal);
```
```suggestion
    async::join_signal_t context_join_signal = m_threadpool.make_join_signal();
    async::join_signal_t data_join_signal    = m_threadpool.make_join_signal();
    async::join_token_t context_join_token   = m_threadpool.get_join_token(context_join_signal);
    async::join_token_t data_join_token      = m_threadpool.get_join_token(data_join_signal);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+            l_data_join_token              = data_join_token
+        ]() mutable -> async::TaskVariant
+        {
+            return chunk_task(l_chunk_request,
```
```suggestion
            return this->chunk_task(l_chunk_request,
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+}
+//-------------------------------------------------------------------------------------------------------------------
+async::TaskVariant AsyncScanContextLegacy::chunk_task(const ChunkRequest &chunk_request,
+    std::shared_future<void> &context_stop_flag,
+    std::shared_future<void> &data_stop_flag,
+    std::shared_ptr<std::promise<sp::scanning::ChunkContext>> &chunk_context_ptr_out,
+    std::shared_ptr<std::promise<sp::scanning::ChunkData>> &chunk_data_ptr_out,
+    async::join_token_t &context_join_token_out,
+    async::join_token_t &data_join_token_out)
```
```suggestion
    std::shared_future<void> &&context_stop_flag,
    std::shared_future<void> &&data_stop_flag,
    std::promise<sp::scanning::ChunkContext> &&chunk_context_ptr,
    std::promise<sp::scanning::ChunkData> &&chunk_data_ptr,
    async::join_token_t &&context_join_token,
    async::join_token_t &&data_join_token)
```

These are all values that should be moved into the task. Also, not sure what those shared_ptr are for.

j-berman replied:
These changes + moving all values into the task end up triggering `error: static assertion failed: std::function target must be copy-constructible` -- suggestion?

UkoeHB replied:
Oh uhh, guess leave it as-is then.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    async::join_token_t &context_join_token_out,
+    async::join_token_t &data_join_token_out)
+{
+    async::fanout_token_t fanout_token{m_threadpool.launch_temporary_worker()};
```
I think either this fanout token needs to be destroyed after the chunk request, or it should only be created farther down the callstack within `handle_chunk_context` when you can guarantee that the chunk request doesn't burn CPU.

I'm guessing this current implementation has some thread contention in the enote finding context when at max capacity. Would be curious to see profiling details for contention, if you can get them.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    }
+}
+//-------------------------------------------------------------------------------------------------------------------
+void AsyncScanContextLegacy::handle_chunk_context(const ChunkRequest &chunk_request,
```
```suggestion
void AsyncScanContextLegacy::handle_chunk_request(const ChunkRequest &chunk_request,
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        const std::size_t chunk_size = sp::scanning::chunk_size(chunk_context);
+        const std::uint64_t gap = requested_chunk_size - chunk_size;
+        if (gap > 0)
+        {
```
```suggestion
        if (gap == 0)
            return;

```
Less nesting

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+            else
+            {
+                // Advance scan index to the start of the gap for next task
+                m_scan_index.store(gap_start_index, std::memory_order_relaxed);
```
```suggestion
                // Pull scan index back to the start of the gap for next task.
                // - For serial scan contexts (when `m_config.pending_chunk_queue_size == 1`),
                //   we can't launch a gap-filler task. Instead, we say the next serial task will start
                //   at the gap start index.
                m_scan_index.store(gap_start_index, std::memory_order_relaxed);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+{
+    // Don't need to launch the next task if found the terminal chunk, we're done!
+    if (!chunk_is_terminal_chunk)
+    {
```
```suggestion
    if chunk_is_terminal_chunk
        return;

```
Less nesting

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+{
+    THROW_WALLET_EXCEPTION_IF(!pending_queue_lock.owns_lock(), tools::error::wallet_internal_error,
+        "Pending queue must be locked to launch the next chunk task");
+
```
```suggestion
bool AsyncScanContextLegacy::try_launch_next_chunk_task(const std::unique_lock<std::mutex> &pending_queue_lock)
{
    THROW_WALLET_EXCEPTION_IF(!pending_queue_lock.owns_lock(), tools::error::wallet_internal_error,
        "Pending queue must be locked to launch the next chunk task");

    if !check_launch_next_task(lock)
        return false;
```
I'd do it like this. Then when `launch_next_chunk_task` is called below it can instead be called directly.

You can also move this debug statement to the top of this function: `MDEBUG("Attempting to launch chunk task at ...`

UkoeHB replied:
Also the `pending_queue_lock.owns_lock(), ` check isn't strictly safe, because it could be *any* lock passed in. You probably need a thread-local id + atomic member variable set to equal the thread id that has the lock (with scoped leave handler that unsets the thread id before dropping the top-level lock guard). Then check the locked thread id equals the current thread's id.

j-berman replied:
> because it could be any lock passed in

It's only ever called with a thread-local unique lock. When I see a unique lock, I understand them to be thread-local ([they're supposed to be scoped to a thread](https://stackoverflow.com/a/40665203)).

I'm just going to remove this check instead of implementing the suggestion if this is a no go as is. It's there to provide some clarity on how the scoped lock is expected to be controlled. If it doesn't help provide that clarity, I'm just going to remove it

UkoeHB replied:
> It's only ever called with a thread-local unique lock. When I see a unique lock, I understand them to be thread-local ([they're supposed to be scoped to a thread](https://stackoverflow.com/a/40665203)).

Right, but there is no guarantee in the check that you locked the right mutex. It's just implied/encouraged by the code structure. Someone reading the code would still need to validate where the lock comes from. So I'd add a comment explaining what the lock is supposed to hold.

j-berman replied:
Implemented [this](https://github.com/UkoeHB/monero/pull/23#discussion_r1578678809) in the latest

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+void AsyncScanContextLegacy::handle_terminal_chunk()
+{
+    // Clear up everything left in the queue
+    wait_until_pending_queue_clears();
```
```suggestion
    this->wait_until_pending_queue_clears();
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        // The chain must have advanced since we started scanning, restart scanning from the highest scan
+        MDEBUG("The chain advanced since we started scanning, restart from last scan");
+        std::unique_lock<std::mutex> lock{m_pending_queue_mutex};
+        start_scanner(m_last_scanned_index, m_max_chunk_size_hint, lock);
```
```suggestion
        this->start_scanner(m_last_scanned_index, m_max_chunk_size_hint, lock);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        // We're good to go, advance the end scan index
+        MDEBUG("We're prepared for the end condition, we scanned to " << m_last_scanned_index);
+        m_end_scan_index = m_last_scanned_index;
+        m_scanner_ready.store(true, std::memory_order_relaxed); // mark the scanner ready for the end condition
```
Setting `m_scanner_ready` here is confusing, since it was just set to `false` in the queue-clear and it seems like `start_scanner` is supposed to be in charge of setting it.

j-berman replied:
Added a new boolean `m_scanner_finished` to capture this state

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        // Wait until all work in the pending queue is done, not just contexts
+        // TODO: wait until every task in the pool has returned
+        m_threadpool.work_while_waiting(clear_chunk.pending_data.data_join_condition,
+            async::DefaultPriorityLevels::MAX);
```
As a first step I'd set the stop signals in the pending chunk context/data here.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+//-------------------------------------------------------------------------------------------------------------------
+void AsyncScanContextLegacy::wait_until_pending_queue_clears()
+{
+    // TODO: implement a clean safe cancel instead of waiting
```
I don't think you can cancel without *some* waiting. The tasks can't be detached since they are bound to `this`.

j-berman replied:
Fair, re-worded the comment.

In addition to your other suggestions, it probably would take a decent amount of work for a future PR, but we could incorporate the ability to cancel in-progress http requests.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    // launch tasks until the queue fills up
+    while (check_launch_next_task(pending_queue_lock))
+    {
+        launch_next_chunk_task(pending_queue_lock);
```
```suggestion
    while (this->check_launch_next_task(pending_queue_lock))
    {
        this->launch_next_chunk_task(pending_queue_lock);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    wait_until_pending_queue_clears();
+
+    std::unique_lock<std::mutex> pending_queue_lock{m_pending_queue_mutex};
+    start_scanner(start_index, max_chunk_size_hint, pending_queue_lock);
```
```suggestion
    this->wait_until_pending_queue_clears();

    std::unique_lock<std::mutex> pending_queue_lock{m_pending_queue_mutex};
    this->start_scanner(start_index, max_chunk_size_hint, pending_queue_lock);
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+        if (oldest_chunk_result == async::TokenQueueResult::QUEUE_EMPTY)
+        {
+            // We should be done scanning now
+            return handle_end_condition();
```
```suggestion
            return this->handle_end_condition();
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    {
+        MDEBUG("Encountered terminal chunk starting at " << oldest_context.start_index
+            << " (expected to start at " << oldest_request.start_index << ")");
+        handle_terminal_chunk();
```
```suggestion
        this->handle_terminal_chunk();
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,797 @@
+    if (m_num_blocks_in_chain > 0)
+        LOG_PRINT_L0("Block " << m_last_scanned_index << " / " << m_num_blocks_in_chain);
+
+    return std::make_unique<sp::scanning::AsyncLedgerChunk>(m_threadpool,
+        std::move(oldest_chunk.pending_context),
+        std::move(pending_chunk_data),
+        std::vector<rct::key>{rct::zero()});
```
I just realized `AsyncLedgerChunk` is unsafe because the threadpool reference can become invalid. I'll have to fix this in a follow-up.

UkoeHB reviewed (changes requested):
Review done

I did not validate all the edge conditions, but overall it looks correct.

j-berman committed:
daemon RPC: default new optional RPC fields to falsey

j-berman committed:
koe comments

j-berman committed:
remove m_scan_end_index and add m_scanner_finished

jeffro256 commented on src/seraphis_main/enote_finding_context.h:
```diff
@@ -84,4 +89,131 @@ class EnoteFindingContextLedger
+    rct::key transaction_id;
+    uint64_t unlock_time;
+    sp::TxExtra tx_memo;
+    uint64_t total_enotes_before_tx;
```
This doesn't capture all the information needed to index legacy enotes. It might be better to make this `std::vector<legacy_output_index_t>`, one for each enote after #42 is merged. Alternatively, it could be a `std::vector<uint64_t>` with an additional `bool  is_rct` field.

j-berman replied:
Ya, it'll need that PR. I have a TODO in `scan_context_async_mock` for it

j-berman reviewed (commented):
Thank you for the detailed review :) Ready for re-review

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -198,88 +212,103 @@ static void rpc_get_blocks_internal(const ChunkRequest &chunk_request,
+
+    prepare_unscanned_legacy_transaction(miner_tx_hash,
+        block.miner_tx,
+        get_total_output_count_before_tx(output_indices[0].indices),
```
`output_indices` size is unchecked in this function. I realize you check it farther up the call stack, but redundancy is good for maintainability.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -198,88 +212,103 @@ static void rpc_get_blocks_internal(const ChunkRequest &chunk_request,
+    unscanned_block_out.block_hash = rct::hash2rct(cryptonote::get_block_hash(block));
+    unscanned_block_out.prev_block_hash = rct::hash2rct(block.prev_id);
+
+    chunk_context_out.block_ids.emplace_back(unscanned_block_out.block_hash);
```
Chunk context block ids should be cleared at the top of this function, since it's an `out` variable. And probably `.reserve` the correct capacity.

j-berman replied:
Technically it should've been an `inout` variable, not supposed to clear in that function. I moved chunk context handling out to the caller for better clarity + added the `.reserve` in the latest

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -657,25 +697,31 @@ std::unique_ptr<sp::scanning::LedgerChunk> AsyncScanContextLegacy::handle_end_co
 //-------------------------------------------------------------------------------------------------------------------
 void AsyncScanContextLegacy::wait_until_pending_queue_clears()
```
I'd rename to `close_and_clear_pending_queue`.

UkoeHB commented on src/seraphis_main/scan_machine.cpp:
```diff
@@ -87,18 +96,20 @@ static std::uint64_t get_reorg_avoidance_depth(const std::uint64_t reorg_avoidan
     // 1. set reorg avoidance depth
     const std::uint64_t reorg_avoidance_depth{
-            get_reorg_avoidance_depth(reorg_avoidance_increment, num_reorg_avoidance_backoffs)
+            get_reorg_avoidance_depth(reorg_avoidance_increment, force_reorg_avoidance_increment, num_reorg_avoidance_backoffs)
```
```suggestion
            get_reorg_avoidance_depth(reorg_avoidance_increment,
                force_reorg_avoidance_increment,
                num_reorg_avoidance_backoffs)
```
Your font size is too small :p

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,241 @@
+    ~AsyncScanContextLegacy()
+    {
+        std::lock_guard<std::mutex> lock{m_async_scan_context_mutex};
+        wait_until_pending_queue_clears();
```
```suggestion
        this->wait_until_pending_queue_clears();
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,838 @@
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static uint64_t get_total_output_count_before_tx(std::vector<uint64_t> output_indices)
```
```suggestion
static uint64_t get_total_output_count_before_tx(const std::vector<uint64_t> &output_indices)
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,838 @@
+    // Older daemons can return more blocks than requested because they did not support a max_block_count req param.
+    // The scanner expects requested_chunk_size blocks however, so we only care about the blocks up until that point.
+    // Note the scanner can also return *fewer* blocks than requested if at chain tip or the chunk exceeded max size.
+    // Warning: fill_gap_if_needed will throw if we don't resize the chunk context before calling it
```
```suggestion
```
Don't think this adds clarity. In this code it's clear `unscanned_chunk_out` needs to be resized in order to insert blocks, it's independent of `fill_gap_if_needed`.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,838 @@
+    const std::uint64_t n_blocks_in_chain = m_num_blocks_in_chain.load(std::memory_order_relaxed);
+    chunk_is_terminal_chunk_out = is_terminal_chunk(chunk_context, n_blocks_in_chain);
+
+    // Use the terminal chunk to update the top block hash if the chunk isn't empty.
+    // - This is required if the daemon RPC did NOT provide the top block hash (e.g. when pointing to an older
+    //   daemon), in which case we have to use the last block ID in the terminal chunk to set the top block hash.
```
What if a reorg replaces blocks without adding to chain length? Just wait until the chain gets longer to detect it?

j-berman replied:
Good spot -- mishandled this circumstance with that round of changes. Check the latest

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,838 @@
+            l_chunk_context                = std::make_shared<std::promise<sp::scanning::ChunkContext>>(std::move(chunk_context_handle)),
+            l_chunk_data                   = std::make_shared<std::promise<sp::scanning::ChunkData>>(std::move(chunk_data_handle)),
+            l_context_join_token           = context_join_token,
+            l_data_join_token              = data_join_token
```
```suggestion
            l_chunk_request      = chunk_request,
            l_context_stop_flag  = context_stop_signal.get_future().share(),
            l_data_stop_flag     = data_stop_signal.get_future().share(),
            l_chunk_context      = std::make_shared<std::promise<sp::scanning::ChunkContext>>(std::move(chunk_context_handle)),
            l_chunk_data         = std::make_shared<std::promise<sp::scanning::ChunkData>>(std::move(chunk_data_handle)),
            l_context_join_token = context_join_token,
            l_data_join_token    = data_join_token
```

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,838 @@
+        clear_chunk.pending_data.stop_signal.set_value();
+
+        // Wait until all work in the pending queue is done, not just contexts
+        // TODO: wait until every task in the pool has returned
```
Drain the queue into a vec, then your aggregate wait function just loops the vec once to check all of the inner join conditions (`join_condition_t` is an alias for `std::function<bool()>`).

You still need the outer loop to check the pool for entries added by previously-removed tasks.

j-berman replied:
I follow this suggestion and will do (also waiting to work until after sending all the stop signals should speed this up).

AFAIU checking the inner join conditions here is essentially checking that the join tokens are set to `nullptr`. But the tasks still can have more work they're doing after setting the join tokens. Example: the last `this->try_launch_next_chunk_task(chunk_is_terminal_chunk);` can execute and attempt to schedule a new task, even after we've drained the pending queue and waited on all inner join conditions. Am I missing something there?

UkoeHB replied:
Yeah checking the join condition needs to synchronize with the end of the task, so you know if all tasks are finished and the queue is empty that the queue can't gain any more tasks.

I added a comment for fixing this.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.h:
```diff
@@ -0,0 +1,241 @@
+    {
+        std::lock_guard<std::mutex> lock{m_async_scan_context_mutex};
+        wait_until_pending_queue_clears();
+    }
```
I'd add a comment explaining the safety model here. It *only* works because all tasks with copies of `this` are tracked in the pending queue. When the pending queue returns empty (after draining and working on all removed tasks), you know that there are no lingering tasks with copies of `this`.

UkoeHB reviewed (changes requested):
Looking good, minor comments

UkoeHB commented:
Once this is merged, please open as many PRs to monero core as possible to reduce the file overlap between this branch and core.

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -0,0 +1,838 @@
+
+    // Finished scanning the chunk
+    chunk_data_ptr_out->set_value(std::move(chunk_data));
+    data_join_token_out = nullptr;
```
Yeah you're right, this should not be set to `nullptr` here, the token should be set by dropping it when you exit.

j-berman committed:
koe comments 2

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -295,22 +293,29 @@ static void parse_rpc_get_blocks(const ChunkRequest &chunk_request,
-        "Pending queue must be locked to check next task launch");
+    THROW_WALLET_EXCEPTION_IF(!m_pending_queue_mutex.thread_owns_lock(), tools::error::wallet_internal_error,
+        "this thread does not own the pending queue mutex");
 
```
Ohh nice idea :)

UkoeHB commented on src/seraphis_mocks/scan_context_async_mock.cpp:
```diff
@@ -697,36 +725,51 @@ void AsyncScanContextLegacy::wait_until_pending_queue_clears()
+    while (!drained_chunks.empty())
+    {
+        auto &clear_chunk = drained_chunks.back();
+        MDEBUG("Waiting to clear onchain chunk starting at " << clear_chunk.chunk_request.start_index);
 
-        // Wait until all work in the pending queue is done, not just contexts
-        // TODO: wait until every task in the pool has returned
+        // Wait until **data** join condition is set, we're not waiting on just the contexts
         m_threadpool.work_while_waiting(clear_chunk.pending_data.data_join_condition,
             async::DefaultPriorityLevels::MAX);
 
-        clear_chunk_result = m_pending_chunk_queue.force_pop(clear_chunk);
+        drained_chunks.pop_back();
     }
```
There can be more tasks in the queue after this because task insertion uses `force_push`. These loops need to be wrapped in another loop that continues until the queue is empty.

j-berman replied:
The only way `force_pop` can return  `async::TokenQueueResult::SHUTTING_DOWN` is if the queue is empty

https://github.com/UkoeHB/monero/blob/c7422607385a070d9e0069f96598be78336f1ef8/src/async/token_queue.h#L109-L110

Since this function is holding `m_async_scan_context_mutex`, `m_scanner_ready` is false meaning no new tasks could be scheduled while the pending tasks are all running to completion, and we've now made sure to work until all pending tasks are done at this point, nothing can theoretically exist that uses `force_push` to insert a task in parallel at this point in this function

j-berman replied:
Writing that comment^ made me re-check the "gap fill" edge case which wasn't checking `m_scanner_ready` before `force_push`. Done. That comment should hold now^

UkoeHB reviewed (changes requested):
One comment remaining

j-berman committed:
don't fill gap if scanner is not ready for use

j-berman committed:
consolidate pushing next chunk task into 1 func

j-berman commented:
Submitted PR's for the simple modifications to the core monero repo for the async scanner. If daemons run those changes, they'll be fully compatible with the async scanner in this PR.
